lib/Parse/Lexer.cpp

```diff
@@ -1884,23 +1884,12 @@ void Lexer::lexStringLiteral(unsigned CustomDelimiterLen) {
         // Successfully scanned the body of the expression literal.
         ++CurPtr;
         continue;
+      } else if (!isMultilineString) {
+        diagnose(CurPtr, diag::lex_unterminated_string);
+        diagnose(CurPtr, diag::string_interpolation_unclosed);
       } else if (*CurPtr == '\r' || *CurPtr == '\n') {
-        if (IsMultilineString) {
-          // The only case we reach here is unterminated single line string in the
-          // interpolation. For better recovery, go on after emitting an error.
-          diagnose(CurPtr, diag::string_interpolation_unclosed);
-          diagnose(--TmpPtr, diag::opening_paren);
-          if (CurPtr != BufferEnd) diagnose(CurPtr, diag::lex_unterminated_string);
-          wasErroneous = true;
-          continue;
-        } else {
-          diagnose(CurPtr, diag::string_interpolation_unclosed);
-          diagnose(--TmpPtr, diag::opening_paren);
-          if (*CurPtr != ')') diagnose(CurPtr, diag::lex_unterminated_string);
-          return formToken(tok::unknown, TokStart);
-        }
-      } else {
-        diagnose(TokStart, diag::lex_unterminated_string);
+        diagnose(CurPtr, diag::lex_unterminated_string);
+        diagnose(CurPtr, diag::string_interpolation_unclosed);
         return formToken(tok::unknown, TokStart);
       }
```

Comment:
Looks like you lost the last `else` statement along the way. We still need it to emit `lex_unterminated_string` (and return) in case none of the previous conditions were met, as a fallback.